### DISCUSSION TITLE
Can Avalonia be used in a way similar to ImGui

### DISCUSSION THREAD

**Discussion by denys-volodarskyi** [AvaloniaUI/Avalonia#13411] [Q&A]
https://github.com/ocornut/imgui

It outputs optimized vertex buffers that you can render anytime in your 3D-pipeline-enabled application. 

Can Avalonia be used this way?

- I inject my avalonia app into address space of the target process (as a dll or exe)
- Run Avalonia, it renders graphics
- Output it via D3D

**Comment by thevortexcloud** [comment]
Not sure. But my understanding is that Avalonia can be embedded in anything that:

1. Supports .NET (although you can probably get around this using AOT, like [this](https://ericsink.com/native_aot/sgbridge_050.html) person who made Rust bindings)
2. Supports SkiaSharp

As such there are people who have embedded Avalonia inside 3d game engines such as Unity, [Unreal](https://pbs.twimg.com/media/FhrBiwSWQAAxQfu?format=jpg&name=large), and [Godot](https://github.com/MrJul/Estragonia).

You technically don't even need "graphics" there is [this](https://github.com/jinek/Consolonia) project which makes Avalonia useable for terminal apps.

**Reply by kyuranger** [reply]
BTW, what does the engine render these games? SkiaSharp or the engine of Unity/Unreal/Goddot itself?

**Reply by thevortexcloud** [reply]
I am not sure. I am not particularly familiar with how Avalonia handles rendering. Although the core team did put up this blog post awhile back that does explain some of it.

https://avaloniaui.net/Blog/avalonia-platform-support-why-it-s-simple

The Godot implementation is open source if you feel like digging through it too. It's a surprisingly small amount of code for it. In which case it _looks_ like it's being rendered to a Skia surface and then renders it to a Godot texture and does some interop with Godot's GPU handling. But I only skimmed it so I likely missed something.
